api/v1alpha1/inferenceset_types.go

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) KAITO authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// InferenceSetTemplate defines the template for creating InferenceSet instances.
+type InferenceSetTemplate struct {
+	Resource  ResourceSpec  `json:"resource,omitempty"`
+	Inference InferenceSpec `json:"inference,omitempty"`
+}
+
+// InferenceSetSpec defines the desired state of InferenceSet
+type InferenceSetSpec struct {
+	// Template is the template used to create the InferenceSet.
+	Template InferenceSetTemplate `json:"template"`
+	// Replicas is the desired number of workspaces to be created.
+	// +optional
+	// +kubebuilder:default:=1
+	Replicas *int `json:"replicas,omitempty"`
+	// NodeCountLimit is the maximum number of GPU nodes that can be created for the InferenceSet.
+	// If not specified, there is no limit on the number of GPU nodes that can be created.
+	// +optional
+	NodeCountLimit *int `json:"nodeCountLimit,omitempty"`
+	// workspace created by InferenceSet controller would use this label in resource.labelSelector
+	// +required
+	Selector *metav1.LabelSelector `json:"labelSelector"`
+	// UpdateStrategy indicates the strategy to use to replace existing workspaces with new ones.
+	// +optional
+	// +kubebuilder:default:={"type":"RollingUpdate","rollingUpdate":{"maxUnavailable":1}}
+	// +kubebuilder:validation:Schemaless
+	// +kubebuilder:pruning:PreserveUnknownFields
+	UpdateStrategy appsv1.StatefulSetUpdateStrategy `json:"updateStrategy,omitempty"`
+}
+
+// InferenceSetStatus defines the observed state of InferenceSet
+type InferenceSetStatus struct {
+	// Conditions report the current conditions of the InferenceSet.
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
```

Comment:
do we need to add other status field for this controller?
like https://github.com/kaito-project/kaito/commit/f46c916d626e801a324868f24adf311ac7e5a279#diff-802d68cb9e30b05ed85f62332cadfe391e3ca842694c657c111c1d454316b045R110

Reply:
ref replicaset.status https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/replica-set-v1/#ReplicaSetStatus

Reply:
I have added `Replicas` and  `Conditions` in `InferenceSetStatus` for now, we could add more during implementation if that's necessary. The InferenceSet controller should be only responsible to create workspace CR and when ensure `.status.TargetNodeCount  * replica <= nodeCountLimit` during creating multiple replicas of workspace CRs.